### DISCUSSION TITLE
Redirect within getServerSideProps

### DISCUSSION THREAD

**Discussion by miikee-us** [vercel/next.js#37429] [Help] [answered]
Looking for a little guidance on this as I'm unable to figure out why the code below does not work. I'm ultimately trying to redirect to a new page if a firebase document already exists via getServerSideProps. If I move the return/redirect outside of the of the docRef.get() then it redirects no problem. In the IF conditional though it will not redirect and I know for a fact that the doc.exists is returning **true**.

```
export const getServerSideProps = withAuthUserTokenSSR({
    whenUnauthed: AuthAction.REDIRECT_TO_LOGIN,
  })(async ({ AuthUser, req }) => {
    try {
        var docRef = firebase.firestore().collection("test-collection").doc("test-document");

        docRef.get().then((doc) => {
            if (doc.exists) {         
                return {
                    redirect: {
                      destination: "/",
                      permanent: false,
                    },
                    props: {},
                  };
            } else {
                return { props: {} };
            }
        }).catch((e) => {
            return { props: {} };
        })
    } catch (e) {
        return { props: {} };
    }
})
export default withAuthUser()(Page)
```

**Comment by icyJoseph** [comment, marked as answer]
Looks like a problem following async code.

I assume `docRef.get` returns a promise, that means that the callback in `then` is done _later_, so the synchronous code evaluation moves on, and nothing is returned. After the synchronous code is done, next.js continues its thing, and then a function (the callback inside then) runs and returns the redirect, but nobody is there to collect it.

Could you please try with:

```ts
export const getServerSideProps = withAuthUserTokenSSR({
  whenUnauthed: AuthAction.REDIRECT_TO_LOGIN,
})(async ({ AuthUser, req }) => {
  try {
    const docRef = firebase
      .firestore()
      .collection("test-collection")
      .doc("test-document");

    const doc = await docRef.get();
    if (doc.exists) {
      return {
        redirect: {
          destination: "/",
          permanent: false,
        },
        props: {},
      };
    } else {
      return { props: {} };
    }
  } catch (e) {
    return { props: {} };
  }
});

export default withAuthUser()(Page);

```

Let me know how it goes, I don't have a test bench setup to resolve your issue, so perhaps a few tweaks are needed. I also can't find the firebase types to make full conclusion.

**Reply by miikee-us** [reply]
Thank you @icyJoseph ! this makes a lot of sense on why it wasn't working. I have tried this out and it seems to be working as expected.
